### PR TITLE
Bump typescript to `4.9.5` and build target to `es2020`

### PR DESCRIPTION
**Pull request by konstantin-msft**
- Bump `typescript` to 4.9.5.
- Bump target/module versions to `es2020` when applicable.
- Replace calls to deprecated `mocked()` from `ts-jest/utils` with `jest.spyOn()`.
- Fix types' coercions to facilitate compatibility with the new ts version.
- Make generic type extend `ServerAuthorizationTokenResponse` for `NetworkManager.sendPostRequest()` in `msal-common` to handle post processing properly.
- Make `serverTelemetryManager.cacheFailedRequest()` accept any types of errors.

### CODE REVIEW

File: lib/msal-common/src/account/AuthToken.ts
Comment (konstantin-msft) [inline lib/msal-common/src/account/AuthToken.ts:47]: Type coercions are prohibited in ts v4.

File: lib/msal-node/src/client/ClientApplication.ts
Comment (konstantin-msft) [inline lib/msal-node/src/client/ClientApplication.ts]: `serverTelemetryManager.cacheFailedRequest(e)` expects an instance of `AuthError`. We can't pass an arbitrary error.

File: lib/msal-node/test/client/ClientAssertion.spec.ts
Comment (konstantin-msft) [inline lib/msal-node/test/client/ClientAssertion.spec.ts:5]: Deprecated

File: lib/msal-common/src/network/NetworkManager.ts
Comment (konstantin-msft) [inline lib/msal-common/src/network/NetworkManager.ts:35]: `T` must extend `ServerAuthorizationTokenResponse` because `ThrottlingUtils.postProcess()` expects it as an input.

File: lib/msal-react/tsconfig.json
Comment (konstantin-msft) [inline lib/msal-react/tsconfig.json]: Required for proper class transpiling.

File: change/msal-1a69cf50-db5e-4aaf-ab86-32f98e28bec4.json
Comment (tnorling) [inline change/msal-1a69cf50-db5e-4aaf-ab86-32f98e28bec4.json:4]: Shouldn't be making any changes to msal-core

File: change/@azure-msal-node-extensions-8fbfdb4b-83ee-4c5f-a036-594a7aeb4f06.json
Comment (tnorling) [inline change/@azure-msal-node-extensions-8fbfdb4b-83ee-4c5f-a036-594a7aeb4f06.json:4]: Changefiles for browser, common and node?

File: change/msal-1a69cf50-db5e-4aaf-ab86-32f98e28bec4.json
Comment (konstantin-msft) [reply change/msal-1a69cf50-db5e-4aaf-ab86-32f98e28bec4.json:4]: May I ask you why not?

File: change/@azure-msal-node-extensions-8fbfdb4b-83ee-4c5f-a036-594a7aeb4f06.json
Comment (konstantin-msft) [reply change/@azure-msal-node-extensions-8fbfdb4b-83ee-4c5f-a036-594a7aeb4f06.json:4]: Beachball root scope for v3 is `["lib/*", "extensions/*"]`. That's why there are no dedicated files for the packages you listed above.
@hectormmg

File: change/msal-1a69cf50-db5e-4aaf-ab86-32f98e28bec4.json
Comment (tnorling) [reply change/msal-1a69cf50-db5e-4aaf-ab86-32f98e28bec4.json:4]: That's MSAL.js v1 which is no longer receiving updates (and this would need to be a major change there)

File: change/@azure-msal-node-extensions-8fbfdb4b-83ee-4c5f-a036-594a7aeb4f06.json
Comment (tnorling) [reply change/@azure-msal-node-extensions-8fbfdb4b-83ee-4c5f-a036-594a7aeb4f06.json:4]: More likely this is happening because the v3 branch already has changefiles and beachball is configured to use dev as the root, not the v3 branch. We should likely update the beachball config in the v3 branch to use the v3 branch as root.

File: change/msal-1a69cf50-db5e-4aaf-ab86-32f98e28bec4.json
Comment (sameerag) [reply change/msal-1a69cf50-db5e-4aaf-ab86-32f98e28bec4.json:4]: If we are supporting security issues (which I recall this is), we may have to push it to V1 too. What is the cost in terms of breaking change?

Comment (tnorling) [reply change/msal-1a69cf50-db5e-4aaf-ab86-32f98e28bec4.json:4]: Let's discuss the security aspect offline. The cost is that we already have an MSAL.js v2, we shouldn't add confusion and the maintenance cost of supporting multiple older versions unless absolutely necessary.

Comment (konstantin-msft) [reply change/msal-1a69cf50-db5e-4aaf-ab86-32f98e28bec4.json:4]: Reverted `msal-core` changes. Thanks for bringing the context Thomas, that's very helpful.

File: lib/msal-browser/src/interaction_client/SilentRefreshClient.ts
Comment (sameerag) [inline lib/msal-browser/src/interaction_client/SilentRefreshClient.ts]: Was there a possibility `e` is not of the type `AuthError`? If NO, we can get rid of the `if` and if `YES`, we still need to cache it outside of the condition right?

Comment (konstantin-msft) [reply lib/msal-browser/src/interaction_client/SilentRefreshClient.ts]: Good catch, thank you. `if (e instanceof AuthError)` is redundant here. Fixed.

File: change/@azure-msal-node-extensions-8fbfdb4b-83ee-4c5f-a036-594a7aeb4f06.json
Comment (konstantin-msft) [reply change/@azure-msal-node-extensions-8fbfdb4b-83ee-4c5f-a036-594a7aeb4f06.json:4]: Beachball doesn't require changefiles for browser, common, etc even after updating the root branch https://github.com/AzureAD/microsoft-authentication-library-for-js/pull/5780.

File: lib/msal-browser/src/interaction_client/PopupClient.ts
Comment (peterzenz) [inline lib/msal-browser/src/interaction_client/PopupClient.ts]: This looks like a functional change and not a change related to the TS/ES intent of this PR. You mention type coercion changes are not supported in the new TS and I'm seeing this pattern throughout this change.  What happened in the original code if we called `.cacheFailedRequest(e)` and it was not of type `AuthError`?  Did the failure cascade out from `cacheFailedRequest` or was it able to coerce whatever error type that was thrown into something usable?  If it's not of type `AuthError` we might still want to cache a failed response (failure is failure...).  I just don't know what other types we might get in this flow.

File: lib/msal-browser/src/utils/BrowserUtils.ts
Comment (peterzenz) [inline lib/msal-browser/src/utils/BrowserUtils.ts:79]: Maybe a comment on why we're adding the ts-ignore here?  Is there a future fix we're holding off on just to unblock folks?

Comment (peterzenz) [reply lib/msal-browser/src/utils/BrowserUtils.ts:79]: As I look at the other ts-ignores around window I'm thrown back to some earlier discussions on web workers not supporting window references.  @sameerag or @tnorling I thought there were some alternates here that might be more friendly.  I'll admit though I'm not up to speed on the new TS/ES changes that might impact here.

File: lib/msal-browser/test/interaction_client/RedirectClient.spec.ts
Comment (peterzenz) [inline lib/msal-browser/test/interaction_client/RedirectClient.spec.ts:9]: Not as terse, but so much more readable.  :)

**Review by peterzenz — APPROVED** [review]
I have the questions I mentioned in teams about the explicit type changes for TS4 and how those resolve.  I'd also like some clarity on the ts-ignores as to what it would take to address those correctly.  If we want this in though now to unblock I'm OK, but please track bugs/work items around these things if they are unresolved.

File: lib/msal-browser/src/interaction_client/PopupClient.ts
Comment (sameerag) [reply lib/msal-browser/src/interaction_client/PopupClient.ts]: Same comment, similar pattern below.

File: lib/msal-browser/test/interaction_client/RedirectClient.spec.ts
Comment (sameerag) [reply lib/msal-browser/test/interaction_client/RedirectClient.spec.ts:9]: At some point I had all files like this. Can we make this a standard?

File: lib/msal-browser/src/utils/BrowserUtils.ts
Comment (sameerag) [reply lib/msal-browser/src/utils/BrowserUtils.ts:79]: Reading [this](https://github.com/microsoft/TypeScript/issues/36048), I get that we are checking if `window.fetch` and `window.headers` are defined, what is the alternative here?  

May be a `try .. catch` where if `window.fetch` isn't defined, we cannot call the `fetchClient` anyway. I am not sure if that actually throws though but worth a try.

File: lib/msal-browser/src/interaction_client/PopupClient.ts
Comment (konstantin-msft) [reply lib/msal-browser/src/interaction_client/PopupClient.ts]: That's a very good point. I have made `serverTelemetryManager.cacheFailedRequest()` to accept any types of errors to make sure we cache all of them.

File: lib/msal-browser/src/utils/BrowserUtils.ts
Comment (konstantin-msft) [reply lib/msal-browser/src/utils/BrowserUtils.ts:79]: Unfortunately, `try .. catch` won't help here.
image

File: change/msal-1a69cf50-db5e-4aaf-ab86-32f98e28bec4.json
Comment (jo-arroyo) [reply change/msal-1a69cf50-db5e-4aaf-ab86-32f98e28bec4.json:4]: If changes reverted, shouldn't this file be deleted?

File: extensions/msal-node-extensions/tsconfig.json
Comment (jo-arroyo) [inline extensions/msal-node-extensions/tsconfig.json]: The others are being set to ES2020. Is there a reason why we're setting this to esnext?

**Review by peterzenz — APPROVED** [review]

File: lib/msal-node/src/client/ClientApplication.ts
Comment (jo-arroyo) [reply lib/msal-node/src/client/ClientApplication.ts]: If `serverTelemetryManager.cacheFailedRequest()` is able to accept any types of error, do these need to be reverted? (and below)

File: extensions/msal-node-extensions/tsconfig.json
Comment (konstantin-msft) [reply extensions/msal-node-extensions/tsconfig.json]: Good catch, thank you. Fixed.

File: lib/msal-node/src/client/ClientApplication.ts
Comment (konstantin-msft) [reply lib/msal-node/src/client/ClientApplication.ts]: I have already reverted it but forgot to resolve the comment. Sorry about that.

Comment (jo-arroyo) [reply lib/msal-node/src/client/ClientApplication.ts]: But `cacheFailedRequest` is still being called inside the `if (e instanceof AuthError)` block

File: change/msal-1a69cf50-db5e-4aaf-ab86-32f98e28bec4.json
Comment (konstantin-msft) [reply change/msal-1a69cf50-db5e-4aaf-ab86-32f98e28bec4.json:4]: We still need this. It's a common file for changes in lib/** modules.

File: lib/msal-node/src/client/ClientApplication.ts
Comment (konstantin-msft) [reply lib/msal-node/src/client/ClientApplication.ts]: Completely forgot about `msal-node` module. Good catch, thank you. Fixed.

**Review by jo-arroyo — APPROVED** [review]
Thanks for addressing comments!

File: change/@azure-msal-node-extensions-8fbfdb4b-83ee-4c5f-a036-594a7aeb4f06.json
Comment (tnorling) [reply change/@azure-msal-node-extensions-8fbfdb4b-83ee-4c5f-a036-594a7aeb4f06.json:4]: That's because the [command](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/112abff69ebb68d4c8a88e86af6d056cf650e492/package.json#L63) specifically references the branch

File: change/msal-1a69cf50-db5e-4aaf-ab86-32f98e28bec4.json
Comment (tnorling) [reply change/msal-1a69cf50-db5e-4aaf-ab86-32f98e28bec4.json:4]: Not sure I understand this... msal == msal-core. Changefile should only be needed if you're making msal-core changes, which we shouldn't be.

**Pull request merged**
